### PR TITLE
Fix bug when `freezing_rv_and_dims` after a model transformation

### PR DESCRIPTION
**Pull request by ricardoV94**
## Description

Also adds optimization to avoid cloning the fgraph when it will be discarded anyway. 
Benchmarked with long models and can cut `clone_model` (the simplest workflow) by 2x

## Related Issue

- [ ] Closes #
- [ ] Related to #

## Checklist

- [ ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [ ] New feature / enhancement
- [x] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc--7296.org.readthedocs.build/en/7296/

### CODE REVIEW

**Comment by ricardoV94** [comment]
I think I broke something in the fgraph stuff, need to investigate

File: pymc/model/fgraph.py [hunk 1]
```diff
@@ -278,12 +278,18 @@ def fgraph_from_model(
     return fgraph, memo
 
 
-def model_from_fgraph(fgraph: FunctionGraph) -> Model:
+def model_from_fgraph(fgraph: FunctionGraph, mutate_fgraph: bool = False) -> Model:
```
Comment (jessegrabowski) [inline pymc/model/fgraph.py:281]: Every time you call the function you use `mutate_fgraph = True`, should that be the default instead? Or are the instances in this PR special cases (and there are many others)

**Review by jessegrabowski — APPROVED** [review]
Looks good. The bug was just related to the shapes of transformed models not being correctly frozen?

File: pymc/model/fgraph.py (hunk 1 above)
Comment (ricardoV94) [reply pymc/model/fgraph.py:281]: I prefer to ask for consent

**Comment by ricardoV94** [comment]
> Looks good. The bug was just related to the shapes of transformed models not being correctly frozen?

Yes, because the shared dim length was not the exact same variable used in the size of the RVs due to an extra cloning of the later. They would still be resized correctly if you updated the dim length but not frozen when you called the associated utility

**Pull request merged**